symbolic-base/test/Tests/Symbolic/Data/Int.hs

```diff
@@ -0,0 +1,187 @@
+import           ZkFold.Symbolic.Data.Int
+import           ZkFold.Symbolic.Data.Ord
+import           ZkFold.Symbolic.Data.UInt                   (UInt (..))
+import           ZkFold.Symbolic.Interpreter                 (Interpreter (Interpreter))
+
+toss :: Natural -> Gen Integer
+toss (P.fromIntegral -> x) = chooseInteger (-x, x)
+
+toss1 :: Natural -> Gen Integer
+toss1 (P.fromIntegral -> x) = do
+    p <- chooseInteger (1, x)
+    n <- chooseInteger (-x, -1)
+    elements [p, n]
+
+tossp :: Natural -> Gen Integer
+tossp (P.fromIntegral -> x) = chooseInteger (1, x)
+
+type AC a = ArithmeticCircuit a U1 U1
+
+evalBool :: forall a . (Arithmetic a, Binary a) => Bool (AC a) -> a
+evalBool (Bool ac) = exec1 ac
+
+evalBoolVec :: forall a . Bool (Interpreter a) -> a
+evalBoolVec (Bool (Interpreter (Par1 v))) = v
+
+execAcInt ::
+  forall a n r . (Arithmetic a, Binary a) =>
+  Int n r (AC a) -> Vector (NumberOfRegisters a (n+1) r) a
+execAcInt (Int (UInt v)) = exec v
+
+execZpInt :: forall a n r . Int n r (Interpreter a) -> Vector (NumberOfRegisters a (n+1) r) a
+execZpInt (Int (UInt (Interpreter v))) = v
+
+type BinaryOp a = a -> a -> a
+
+type UBinary n b r = BinaryOp (Int n b r)
+
+isHom
+    :: forall n p r
+    .  KnownNat n
+    => PrimeField (Zp p)
+    => KnownRegisterSize r
+    => UBinary n r (Interpreter (Zp p))
+    -> UBinary n r (AC (Zp p))
+    -> BinaryOp Integer
+    -> Integer
+    -> Integer
+    -> Property
+isHom f g h x y = execAcInt (fromConstant x `g` fromConstant y) === execZpInt (fromConstant x `f` fromConstant y)
+              .&. execZpInt (fromConstant x `f` fromConstant y) === execZpInt @(Zp p) @n @r (fromConstant $ x `h` y)
+
+-- with2n :: forall n {r}. KnownNat n => (KnownNat (2 * n) => r) -> r
+-- with2n = withDict (timesNat @2 @n)
+
+specInt'
+    :: forall p n r rs
+    .  PrimeField (Zp p)
+    => KnownNat n
+    => KnownNat (n+1)
+    => KnownRegisterSize rs
+    => r ~ NumberOfRegisters (Zp p) (n+1) rs
+    => KnownNat r
+    => Spec
+specInt' = do
+    let n = value @n
+        m = 2 ^ (11 :: Natural)
+    describe ("Int" ++ show n ++ " specification") $ do
+        it "Zp embeds Integer" $ do
+            x <- toss m
+            return $ toConstant @(Int n rs (Interpreter (Zp p))) (fromConstant x) === x
+        it "Integer embeds Zp" $ \(x :: Int n rs (Interpreter (Zp p))) ->
+            fromConstant (toConstant x) === x
+        it "IsNegative correct" $ do
+            x <- tossp m
+            return $ evalBoolVec @(Zp p) (isNegative @n @rs $ fromConstant x) === 0 .&.
+                     evalBoolVec @(Zp p) (isNegative @n @rs $ fromConstant (- x)) === 1
+        it "IsNegative correct-2" $ do
+            x <- toss m
+            let t = evalBoolVec @(Zp p) (isNegative @n @rs $ fromConstant x)
+            return $ (t === 0) .||. (t === 1)
+        it "Abs correct" $ do
+            x <- tossp m
+            return $ toConstant (abs (fromConstant (-x) :: Int n rs (Interpreter (Zp p)))) === x
```

Comment:
I think for a sanity check we can also check here that `abs x == x` and `abs 0 == 0`